fastembed/splitter.py

```diff
@@ -0,0 +1,206 @@
+
+    @abstractmethod
+    def split_text(self, text: str) -> List[str]:
+        """Split text into multiple components."""
+
+    def _join_docs(self, docs: List[str], separator: str) -> Optional[str]:
+        text = separator.join(docs)
+        if self._strip_whitespace:
+            text = text.strip()
+        if text == "":
+            return None
+        else:
+            return text
+
+    def _merge_splits(
+        self,
+        splits: Iterable[str],
+        separator: str,
+        chunk_size: Optional[int] = None,
+        chunk_overlap: Optional[int] = None,
+    ) -> List[str]:
+        # We now want to combine these smaller pieces into medium size
+        # chunks to send to the LLM.
+        separator_len = self._length_function(separator)
+
+        chunk_size = chunk_size or self._chunk_size
+        chunk_overlap = chunk_overlap or self._chunk_overlap
+
+        docs = []
+        current_doc: List[str] = []
+        total = 0
+        for d in splits:
+            _len = self._length_function(d)
+            if total + _len + (separator_len if len(current_doc) > 0 else 0) > chunk_size:
+                if total > chunk_size:
+                    logger.warning(
+                        f"Created a chunk of size {total}, " f"which is longer than the specified {chunk_size}"
+                    )
+                if len(current_doc) > 0:
+                    doc = self._join_docs(current_doc, separator)
+                    if doc is not None:
+                        docs.append(doc)
+                    # Keep on popping if:
+                    # - we have a larger chunk than in the chunk overlap
+                    # - or if we still have any chunks and the length is long
+                    while total > chunk_overlap or (
+                        total + _len + (separator_len if len(current_doc) > 0 else 0) > chunk_size and total > 0
+                    ):
+                        total -= self._length_function(current_doc[0]) + (separator_len if len(current_doc) > 1 else 0)
+                        current_doc = current_doc[1:]
+            current_doc.append(d)
+            total += _len + (separator_len if len(current_doc) > 1 else 0)
+        doc = self._join_docs(current_doc, separator)
+        if doc is not None:
+            docs.append(doc)
+        return docs
+
+
+class FastEmbedRecursiveSplitter(TextSplitter):
+    """Splitting text by recursively looking at characters.
+
+    Recursively tries to split by different characters with
+    the tokenizer encoding to restrict length.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        chunk_size: int = 4000,
+        chunk_overlap: int = 200,
+        strip_whitespace: bool = True,
+        separators: Optional[List[str]] = None,
+        keep_separator: bool = True,
+        is_separator_regex: bool = False,
+    ) -> None:
+        """Create a new TextSplitter."""
+
+        if not isinstance(tokenizer, Tokenizer):
```

Comment:
We should check that the maximum chunk_size is never more than the max length allowed by the tokenizer. E.g. 512 tokens for BAAI, which can be 2K for some other Tokenizer. Please add that to the docstring and ValueError on that as well.

Reply:
Never mind, I've added this to the validator. Please add a test for this i.e. that the error is indeed raised — and then mark this conversation as resolved — this is the same as previous comment. Can resolve both together.